// @fluentui/react/lib-amd/index.js
define(["require", "exports", "tslib", "./ActivityItem", "./Autofill", "./Announced", "./Breadcrumb", "./Button", "./ButtonGrid", "./Calendar", "./Callout", "./Check", "./Checkbox", "./ChoiceGroup", "./Coachmark", "./Color", "./ColorPicker", "./ComboBox", "./CommandBar", "./ContextualMenu", "./DatePicker", "./DateTimeUtilities", "./DetailsList", "./Dialog", "./Divider", "./DocumentCard", "./DragDrop", "./Dropdown", "./ExtendedPicker", "./Fabric", "./Facepile", "./FloatingPicker", "./FocusTrapZone", "./FocusZone", "./Grid", "./GroupedList", "./HoverCard", "./Icon", "./Icons", "./Image", "./Keytips", "./Keytip", "./KeytipData", "./KeytipLayer", "./Label", "./Layer", "./Link", "./List", "./MarqueeSelection", "./MessageBar", "./Modal", "./Nav", "./OverflowSet", "./Overlay", "./Panel", "./Persona", "./PersonaCoin", "./Pickers", "./Pivot", "./Popup", "./Positioning", "./PositioningContainer", "./ProgressIndicator", "./Rating", "./ResizeGroup", "./ResponsiveMode", "./ScrollablePane", "./SearchBox", "./SelectableOption", "./SelectedItemsList", "./Selection", "./Separator", "./Shimmer", "./ShimmeredDetailsList", "./Slider", "./SpinButton", "./Spinner", "./Stack", "./Sticky", "./Styling", "./SwatchColorPicker", "./TeachingBubble", "./Text", "./TextField", "./ThemeGenerator", "./TimePicker", "./Toggle", "./Tooltip", "./Utilities", "./WeeklyDayPicker", "./WindowProvider", "./utilities/ThemeProvider/index", "./Theme", "./version"], function (require, exports, tslib_1, ActivityItem_1, Autofill_1, Announced_1, Breadcrumb_1, Button_1, ButtonGrid_1, Calendar_1, Callout_1, Check_1, Checkbox_1, ChoiceGroup_1, Coachmark_1, Color_1, ColorPicker_1, ComboBox_1, CommandBar_1, ContextualMenu_1, DatePicker_1, DateTimeUtilities_1, DetailsList_1, Dialog_1, Divider_1, DocumentCard_1, DragDrop_1, Dropdown_1, ExtendedPicker_1, Fabric_1, Facepile_1, FloatingPicker_1, FocusTrapZone_1, FocusZone_1, Grid_1, GroupedList_1, HoverCard_1, Icon_1, Icons_1, Image_1, Keytips_1, Keytip_1, KeytipData_1, KeytipLayer_1, Label_1, Layer_1, Link_1, List_1, MarqueeSelection_1, MessageBar_1, Modal_1, Nav_1, OverflowSet_1, Overlay_1, Panel_1, Persona_1, PersonaCoin_1, Pickers_1, Pivot_1, Popup_1, Positioning_1, PositioningContainer_1, ProgressIndicator_1, Rating_1, ResizeGroup_1, ResponsiveMode_1, ScrollablePane_1, SearchBox_1, SelectableOption_1, SelectedItemsList_1, Selection_1, Separator_1, Shimmer_1, ShimmeredDetailsList_1, Slider_1, SpinButton_1, Spinner_1, Stack_1, Sticky_1, Styling_1, SwatchColorPicker_1, TeachingBubble_1, Text_1, TextField_1, ThemeGenerator_1, TimePicker_1, Toggle_1, Tooltip_1, Utilities_1, WeeklyDayPicker_1, WindowProvider_1, index_1, Theme_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.SharedColors = exports.NeutralColors = exports.MotionAnimations = exports.MotionTimings = exports.MotionDurations = exports.mergeThemes = exports.LocalizedFontNames = exports.LocalizedFontFamilies = exports.FluentTheme = exports.Depths = exports.DefaultSpacing = exports.CommunicationColors = void 0;
    tslib_1.__exportStar(ActivityItem_1, exports);
    tslib_1.__exportStar(Autofill_1, exports);
    tslib_1.__exportStar(Announced_1, exports);
    tslib_1.__exportStar(Breadcrumb_1, exports);
    tslib_1.__exportStar(Button_1, exports);
    tslib_1.__exportStar(ButtonGrid_1, exports);
    tslib_1.__exportStar(Calendar_1, exports);
    tslib_1.__exportStar(Callout_1, exports);
    tslib_1.__exportStar(Check_1, exports);
    tslib_1.__exportStar(Checkbox_1, exports);
    tslib_1.__exportStar(ChoiceGroup_1, exports);
    // export * from './ChoiceGroupOption'; // exported by ChoiceGroup
    tslib_1.__exportStar(Coachmark_1, exports);
    tslib_1.__exportStar(Color_1, exports);
    tslib_1.__exportStar(ColorPicker_1, exports);
    tslib_1.__exportStar(ComboBox_1, exports);
    tslib_1.__exportStar(CommandBar_1, exports);
    tslib_1.__exportStar(ContextualMenu_1, exports);
    tslib_1.__exportStar(DatePicker_1, exports);
    tslib_1.__exportStar(DateTimeUtilities_1, exports);
    tslib_1.__exportStar(DetailsList_1, exports);
    tslib_1.__exportStar(Dialog_1, exports);
    tslib_1.__exportStar(Divider_1, exports);
    tslib_1.__exportStar(DocumentCard_1, exports);
    tslib_1.__exportStar(DragDrop_1, exports);
    tslib_1.__exportStar(Dropdown_1, exports);
    tslib_1.__exportStar(ExtendedPicker_1, exports);
    tslib_1.__exportStar(Fabric_1, exports);
    tslib_1.__exportStar(Facepile_1, exports);
    tslib_1.__exportStar(FloatingPicker_1, exports);
    tslib_1.__exportStar(FocusTrapZone_1, exports);
    tslib_1.__exportStar(FocusZone_1, exports);
    tslib_1.__exportStar(Grid_1, exports);
    tslib_1.__exportStar(GroupedList_1, exports);
    tslib_1.__exportStar(HoverCard_1, exports);
    tslib_1.__exportStar(Icon_1, exports);
    tslib_1.__exportStar(Icons_1, exports);
    tslib_1.__exportStar(Image_1, exports);
    tslib_1.__exportStar(Keytips_1, exports);
    tslib_1.__exportStar(Keytip_1, exports);
    tslib_1.__exportStar(KeytipData_1, exports);
    tslib_1.__exportStar(KeytipLayer_1, exports);
    tslib_1.__exportStar(Label_1, exports);
    tslib_1.__exportStar(Layer_1, exports);
    tslib_1.__exportStar(Link_1, exports);
    tslib_1.__exportStar(List_1, exports);
    tslib_1.__exportStar(MarqueeSelection_1, exports);
    tslib_1.__exportStar(MessageBar_1, exports);
    tslib_1.__exportStar(Modal_1, exports);
    tslib_1.__exportStar(Nav_1, exports);
    tslib_1.__exportStar(OverflowSet_1, exports);
    tslib_1.__exportStar(Overlay_1, exports);
    tslib_1.__exportStar(Panel_1, exports);
    tslib_1.__exportStar(Persona_1, exports);
    tslib_1.__exportStar(PersonaCoin_1, exports);
    // export * from './PersonaPresence'; (Exported as part of Persona)
    tslib_1.__exportStar(Pickers_1, exports);
    tslib_1.__exportStar(Pivot_1, exports);
    tslib_1.__exportStar(Popup_1, exports);
    tslib_1.__exportStar(Positioning_1, exports);
    tslib_1.__exportStar(PositioningContainer_1, exports);
    tslib_1.__exportStar(ProgressIndicator_1, exports);
    tslib_1.__exportStar(Rating_1, exports);
    tslib_1.__exportStar(ResizeGroup_1, exports);
    tslib_1.__exportStar(ResponsiveMode_1, exports);
    tslib_1.__exportStar(ScrollablePane_1, exports);
    tslib_1.__exportStar(SearchBox_1, exports);
    tslib_1.__exportStar(SelectableOption_1, exports);
    tslib_1.__exportStar(SelectedItemsList_1, exports);
    tslib_1.__exportStar(Selection_1, exports);
    tslib_1.__exportStar(Separator_1, exports);
    tslib_1.__exportStar(Shimmer_1, exports);
    tslib_1.__exportStar(ShimmeredDetailsList_1, exports);
    tslib_1.__exportStar(Slider_1, exports);
    tslib_1.__exportStar(SpinButton_1, exports);
    tslib_1.__exportStar(Spinner_1, exports);
    tslib_1.__exportStar(Stack_1, exports);
    tslib_1.__exportStar(Sticky_1, exports);
    tslib_1.__exportStar(Styling_1, exports);
    tslib_1.__exportStar(SwatchColorPicker_1, exports);
    tslib_1.__exportStar(TeachingBubble_1, exports);
    tslib_1.__exportStar(Text_1, exports);
    tslib_1.__exportStar(TextField_1, exports);
    tslib_1.__exportStar(ThemeGenerator_1, exports);
    tslib_1.__exportStar(TimePicker_1, exports);
    tslib_1.__exportStar(Toggle_1, exports);
    tslib_1.__exportStar(Tooltip_1, exports);
    tslib_1.__exportStar(Utilities_1, exports);
    tslib_1.__exportStar(WeeklyDayPicker_1, exports);
    tslib_1.__exportStar(WindowProvider_1, exports);
    /**
     * Now explicitly declaring Theme exports that are NOT already being exported from Styles.
     * Styles and Theme both exported the same names which causes conflicting
     * star exports with webpack5. See here: https://github.com/microsoft/fluentui/issues/21601.
     */
    tslib_1.__exportStar(index_1, exports);
    Object.defineProperty(exports, "CommunicationColors", { enumerable: true, get: function () { return Theme_1.CommunicationColors; } });
    Object.defineProperty(exports, "DefaultSpacing", { enumerable: true, get: function () { return Theme_1.DefaultSpacing; } });
    Object.defineProperty(exports, "Depths", { enumerable: true, get: function () { return Theme_1.Depths; } });
    Object.defineProperty(exports, "FluentTheme", { enumerable: true, get: function () { return Theme_1.FluentTheme; } });
    Object.defineProperty(exports, "LocalizedFontFamilies", { enumerable: true, get: function () { return Theme_1.LocalizedFontFamilies; } });
    Object.defineProperty(exports, "LocalizedFontNames", { enumerable: true, get: function () { return Theme_1.LocalizedFontNames; } });
    Object.defineProperty(exports, "mergeThemes", { enumerable: true, get: function () { return Theme_1.mergeThemes; } });
    Object.defineProperty(exports, "MotionDurations", { enumerable: true, get: function () { return Theme_1.MotionDurations; } });
    Object.defineProperty(exports, "MotionTimings", { enumerable: true, get: function () { return Theme_1.MotionTimings; } });
    Object.defineProperty(exports, "MotionAnimations", { enumerable: true, get: function () { return Theme_1.MotionAnimations; } });
    Object.defineProperty(exports, "NeutralColors", { enumerable: true, get: function () { return Theme_1.NeutralColors; } });
    Object.defineProperty(exports, "SharedColors", { enumerable: true, get: function () { return Theme_1.SharedColors; } });
});
