// office-addin-dev-certs/lib/defaults.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.usageDataObject = exports.state = exports.locality = exports.domain = exports.daysUntilCertificateExpires = exports.countryCode = exports.certificateName = exports.localhostKeyPath = exports.localhostKeyFileName = exports.localhostCertificatePath = exports.localhostCertificateFileName = exports.caCertificatePath = exports.caCertificateFileName = exports.certificateDirectory = exports.certificateDirectoryName = void 0;
const os = require("os");
const path = require("path");
const usageData = require("office-addin-usage-data");
// Default certificate names
exports.certificateDirectoryName = ".office-addin-dev-certs";
exports.certificateDirectory = path.join(os.homedir(), exports.certificateDirectoryName);
exports.caCertificateFileName = "ca.crt";
exports.caCertificatePath = path.join(exports.certificateDirectory, exports.caCertificateFileName);
exports.localhostCertificateFileName = "localhost.crt";
exports.localhostCertificatePath = path.join(exports.certificateDirectory, exports.localhostCertificateFileName);
exports.localhostKeyFileName = "localhost.key";
exports.localhostKeyPath = path.join(exports.certificateDirectory, exports.localhostKeyFileName);
// Default certificate details
exports.certificateName = "Developer CA for Microsoft Office Add-ins";
exports.countryCode = "US";
exports.daysUntilCertificateExpires = 30;
exports.domain = ["127.0.0.1", "localhost"];
exports.locality = "Redmond";
exports.state = "WA";
// Usage data defaults
exports.usageDataObject = new usageData.OfficeAddinUsageData({
    projectName: "office-addin-dev-certs",
    instrumentationKey: usageData.instrumentationKeyForOfficeAddinCLITools,
    raisePrompt: false,
});
