// office-addin-lint/lib/defaults.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.usageDataObject = exports.PrettierExitCode = exports.ESLintExitCode = exports.lintFiles = void 0;
const usageData = require("office-addin-usage-data");
exports.lintFiles = "src/**/*.{ts,tsx,js,jsx}";
var ESLintExitCode;
(function (ESLintExitCode) {
    ESLintExitCode[ESLintExitCode["NoLintErrors"] = 0] = "NoLintErrors";
    ESLintExitCode[ESLintExitCode["HasLintError"] = 1] = "HasLintError";
    ESLintExitCode[ESLintExitCode["CommandFailed"] = 2] = "CommandFailed";
})(ESLintExitCode = exports.ESLintExitCode || (exports.ESLintExitCode = {}));
var PrettierExitCode;
(function (PrettierExitCode) {
    PrettierExitCode[PrettierExitCode["NoFormattingProblems"] = 0] = "NoFormattingProblems";
    PrettierExitCode[PrettierExitCode["HasFormattingProblem"] = 1] = "HasFormattingProblem";
    PrettierExitCode[PrettierExitCode["CommandFailed"] = 2] = "CommandFailed";
})(PrettierExitCode = exports.PrettierExitCode || (exports.PrettierExitCode = {}));
// Usage data defaults
exports.usageDataObject = new usageData.OfficeAddinUsageData({
    projectName: "office-addin-lint",
    instrumentationKey: usageData.instrumentationKeyForOfficeAddinCLITools,
    raisePrompt: false,
});
